<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<title>Introduction to SpaTopic • SpaTopic</title>
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet">
<link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet">
<script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Introduction to SpaTopic">
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-light" data-bs-theme="light" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">SpaTopic</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">1.2.0.9000</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="active nav-item"><a class="nav-link" href="../articles/SpaTopic.html">Get started</a></li>
<li class="nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles">
<li><a class="dropdown-item" href="../articles/Intro_SpaTopic.html">SpaTopic Basics</a></li>
    <li><a class="dropdown-item" href="../articles/Model_Selection.html">Model Selection</a></li>
    <li><a class="dropdown-item" href="../articles/spleen_analysis.html">Analysis across Multiple Images</a></li>
  </ul>
</li>
<li class="nav-item"><a class="nav-link" href="../news/index.html">Changelog</a></li>
      </ul>
<ul class="navbar-nav">
<li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json">
</form></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/xiyupeng/SpaTopic/" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
      </ul>
</div>


  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">

      <h1>Introduction to SpaTopic</h1>
                        <h4 data-toc-skip class="author">Xiyu Peng</h4>
            
            <h4 data-toc-skip class="date">2025-04-17</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/xiyupeng/SpaTopic/blob/HEAD/vignettes/articles/SpaTopic.Rmd" class="external-link"><code>vignettes/articles/SpaTopic.Rmd</code></a></small>
      <div class="d-none name"><code>SpaTopic.Rmd</code></div>
    </div>

    
    
<div class="section level2">
<h2 id="introduction">Introduction<a class="anchor" aria-label="anchor" href="#introduction"></a>
</h2>
<p>Recent advancements in multiplexed tissue imaging allow for
examination of tissue microenvironments in great detail. These
cutting-edge technologies offer invaluable insights into cellular
heterogeneity and spatial architectures, playing a crucial role in
decoding mechanisms of treatment response and disease progression.
However, gaining a deep understanding of complex spatial patterns
remains challenging.</p>
<p>We present <code>SpaTopic</code>, which implements a novel spatial
topic model to integrate both <strong>cell type</strong> and
<strong>spatial information</strong> to identify the complex spatial
tissue structures without human intervention. The method was originally
developed for image segmentation in computer vision, incorporating
spatial information into the flexible design of regions (image
partitions, analogous to documents in language modeling). We further
adapted and refined the approach to address unique challenges in
cellular images and provide an efficient C++ implementation of the
algorithm in this R package.</p>
<p>Contrasting to computationally intensive K-nearest-neighbor-based
cell neighborhood analysis approaches (such as KNN-kmeans, the default
neighborhood analysis in Seurat v5 R package), <code>SpaTopic</code> is
more scalable to large-scale image datasets without extracting
neighborhood information for every single cell.</p>
<p><code>SpaTopic</code> can be applied either on a
<strong>single</strong> image or across <strong>multiple</strong>
images.</p>
</div>
<div class="section level2">
<h2 id="set-up">Set-up<a class="anchor" aria-label="anchor" href="#set-up"></a>
</h2>
<p>We use a non-small cell lung cancer image to illustrate how to use
<code>SpaTopic</code>. The data object here can be download from <a href="https://drive.google.com/drive/folders/1_mJUjzQXWgUZlwUaLq0HKxX-aqgiQ8eD?usp=sharing" class="external-link">here</a>,
with original public resources available on the <a href="https://nanostring.com/products/cosmx-spatial-molecular-imager/ffpe-dataset/nsclc-ffpe-dataset/" class="external-link">nanostring
website</a>. These images were generated using a 960-plex CoxMx RNA
panel on the Nanostring CoxMx Spatial Molecular Imager platform. We
selected Lung5-1 sample and annotated cells using Azimuth based on the
human lung reference v1.0. The Lung5-1 sample contains 38 annotated cell
types. Since we used healthy lung tissue as the reference, tumor cells
were labeled as ’basal’ cells. Here we mainly followed the cell
annoration step from Satija Lab, with more information can be found <a href="https://satijalab.org/seurat/articles/seurat5_spatial_vignette_2.html#human-lung-nanostring-cosmx-spatial-molecular-imager" class="external-link">here</a>.</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## We use Seurat v5 package to visualize the results.</span></span>
<span><span class="co">## If you still use Seurat v4, you will have the error</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://satijalab.org/seurat" class="external-link">Seurat</a></span>, quietly <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span>;<span class="fu"><a href="https://rdrr.io/r/utils/packageDescription.html" class="external-link">packageVersion</a></span><span class="op">(</span><span class="st">"Seurat"</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] '5.3.0'</span></span>
<span><span class="co">## Load the Seurat object for the image</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/load.html" class="external-link">load</a></span><span class="op">(</span><span class="st">"~/Documents/Research/github/SpaTopic_data/nanostring_example.rdata"</span><span class="op">)</span></span>
<span><span class="co">## for large dataset</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/options.html" class="external-link">options</a></span><span class="op">(</span>future.globals.maxSize <span class="op">=</span> <span class="fl">1e9</span><span class="op">)</span></span></code></pre></div>
<p>We can use the Seurat function <code>ImageDimPlot</code> to visualize
the distribution of cell types on the image.</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://ggplot2.tidyverse.org" class="external-link">ggplot2</a></span><span class="op">)</span></span>
<span><span class="va">celltype.plot</span> <span class="op">&lt;-</span><span class="fu"><a href="https://satijalab.org/seurat/reference/ImageDimPlot.html" class="external-link">ImageDimPlot</a></span><span class="op">(</span><span class="va">nano.obj</span>, fov <span class="op">=</span> <span class="st">"lung5.rep1"</span>, axes <span class="op">=</span> <span class="cn">TRUE</span>, cols <span class="op">=</span> <span class="st">"glasbey"</span>,dark.background <span class="op">=</span> <span class="cn">T</span><span class="op">)</span></span>
<span><span class="va">celltype.plot</span><span class="op">+</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html" class="external-link">theme</a></span><span class="op">(</span>legend.position <span class="op">=</span> <span class="st">"bottom"</span>,legend.direction <span class="op">=</span> <span class="st">"vertical"</span><span class="op">)</span></span></code></pre></div>
<p><img src="SpaTopic_files/figure-html/celltype-1.png" width="672"></p>
</div>
<div class="section level2">
<h2 id="topic-inference-on-a-single-image">Topic Inference on a Single Image<a class="anchor" aria-label="anchor" href="#topic-inference-on-a-single-image"></a>
</h2>
<p>Now, our data is ready. Below we show an example how to use
<code>SpaTopic</code> to identify tissue architectures from multiplexed
images.</p>
<div class="section level3">
<h3 id="input">Input<a class="anchor" aria-label="anchor" href="#input"></a>
</h3>
<p>The required input of SpaTopic is a data frame containing cells
within on a single image or a list of data frames for multiple images.
Each data frame consists of four columns: The image ID, X, Y cell
coordinates, and cell type.</p>
<p>You may use the function <code><a href="../reference/Seurat5obj_to_SpaTopic.html">Seurat5obj_to_SpaTopic()</a></code> to
extract input data from a typical Seurat v5 object. The column name for
cell type information need to be provided via option
<code>group.by</code>.</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/xiyupeng/SpaTopic" class="external-link">SpaTopic</a></span><span class="op">)</span>;<span class="fu"><a href="https://rdrr.io/r/utils/packageDescription.html" class="external-link">packageVersion</a></span><span class="op">(</span><span class="st">"SpaTopic"</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] '1.2.0.9000'</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://r-spatial.github.io/sf/" class="external-link">sf</a></span><span class="op">)</span></span>
<span><span class="co">## Prepare input from Seurat Object</span></span>
<span><span class="va">dataset</span><span class="op">&lt;-</span><span class="fu"><a href="../reference/Seurat5obj_to_SpaTopic.html">Seurat5obj_to_SpaTopic</a></span><span class="op">(</span>object <span class="op">=</span> <span class="va">nano.obj</span>, group.by <span class="op">=</span> <span class="st">"predicted.annotation.l1"</span>,image <span class="op">=</span> <span class="st">"image1"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">dataset</span><span class="op">)</span></span>
<span><span class="co">#&gt;      image        X        Y           type</span></span>
<span><span class="co">#&gt; 1_1 image1 4215.889 158847.7      Dendritic</span></span>
<span><span class="co">#&gt; 2_1 image1 6092.889 158834.7     Macrophage</span></span>
<span><span class="co">#&gt; 3_1 image1 7214.889 158843.7 Neuroendocrine</span></span>
<span><span class="co">#&gt; 4_1 image1 7418.889 158813.7     Macrophage</span></span>
<span><span class="co">#&gt; 5_1 image1 7446.889 158845.7     Macrophage</span></span>
<span><span class="co">#&gt; 6_1 image1 3254.889 158838.7          CD4 T</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="gibbs-sampling">Gibbs Sampling<a class="anchor" aria-label="anchor" href="#gibbs-sampling"></a>
</h3>
<p>This step takes around 60~90 seconds on a regular laptop</p>
<p><strong>Important Notes for Parameter Selection:</strong></p>
<p>When running SpaTopic, both <code>sigma</code> and
<code>region_radius</code> should be set based on image resolution and
tissue complexity:</p>
<ul>
<li><p>For whole-slide imaging applications, select
<code>region_radius</code> to include at least 100 cells per region on
average. Note that different imaging platforms may report spatial
coordinates in either pixels or microns, so adjust parameters
accordingly.</p></li>
<li><p>The <code>sigma</code> parameter should be tuned in conjunction
with <code>region_radius</code>. Empirically, we’ve found that setting
<code>sigma</code> to approximately the square root of
<code>region_radius</code> works well as a starting point for parameter
tuning.</p></li>
</ul>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## Gibbs sampling for SpaTopic</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/system.time.html" class="external-link">system.time</a></span><span class="op">(</span><span class="va">gibbs.res</span><span class="op">&lt;-</span><span class="fu"><a href="../reference/SpaTopic_inference.html">SpaTopic_inference</a></span><span class="op">(</span><span class="va">dataset</span>, ntopics <span class="op">=</span> <span class="fl">7</span>, sigma <span class="op">=</span> <span class="fl">50</span>, region_radius <span class="op">=</span> <span class="fl">400</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">#&gt; [15:48:55] [SpaTopic INFO] Number of cells per image: 100149</span></span>
<span><span class="co">#&gt; [15:48:55] [SpaTopic INFO] Start initialization...</span></span>
<span><span class="co">#&gt; [15:48:55] [SpaTopic INFO] Number of Initializations: 10</span></span>
<span><span class="co">#&gt; [15:49:14] [SpaTopic RESULT] Min perplexity during initialization: 11.6302</span></span>
<span><span class="co">#&gt; [15:49:14] [SpaTopic INFO] Number of region centers selected: 971</span></span>
<span><span class="co">#&gt; [15:49:14] [SpaTopic INFO] Average number of cells per region: 103.14</span></span>
<span><span class="co">#&gt; [15:49:14] [SpaTopic PROGRESS] Initialization complete. Starting Gibbs sampling...</span></span>
<span><span class="co">#&gt; [15:49:50] [SpaTopic COMPLETE] Gibbs sampling completed successfully</span></span>
<span><span class="co">#&gt; [15:49:50] [SpaTopic RESULT] Final model perplexity: 11.3156</span></span>
<span><span class="co">#&gt;    user  system elapsed </span></span>
<span><span class="co">#&gt;  54.815   0.057  54.936</span></span></code></pre></div>
<p>Print the result</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="va">gibbs.res</span><span class="op">)</span></span>
<span><span class="co">#&gt; SpaTopic Results</span></span>
<span><span class="co">#&gt; ----------------</span></span>
<span><span class="co">#&gt; Number of topics: 7 </span></span>
<span><span class="co">#&gt; Perplexity: 11.31563 </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Topic Content(Topic distribution across cell types):</span></span>
<span><span class="co">#&gt;                                 topic1       topic2       topic3      topic4</span></span>
<span><span class="co">#&gt; Alveolar Epithelial Type 1 0.035870295 6.511503e-03 4.541367e-06 0.026643327</span></span>
<span><span class="co">#&gt; Alveolar Epithelial Type 2 0.025386476 3.553900e-02 4.541367e-06 0.017427665</span></span>
<span><span class="co">#&gt; Artery                     0.007545591 2.624548e-06 9.128148e-04 0.001856373</span></span>
<span><span class="co">#&gt; B                          0.018581190 5.800251e-04 3.446035e-01 0.015203195</span></span>
<span><span class="co">#&gt; Basal                      0.025846292 7.753466e-01 1.730261e-03 0.089193312</span></span>
<span><span class="co">#&gt;                                  topic5       topic6      topic7</span></span>
<span><span class="co">#&gt; Alveolar Epithelial Type 1 2.987411e-06 6.348481e-06 0.005341969</span></span>
<span><span class="co">#&gt; Alveolar Epithelial Type 2 2.987411e-06 6.348481e-06 0.006994451</span></span>
<span><span class="co">#&gt; Artery                     5.320579e-03 2.044846e-02 0.006041096</span></span>
<span><span class="co">#&gt; B                          1.912242e-02 3.434528e-03 0.018434717</span></span>
<span><span class="co">#&gt; Basal                      4.902342e-03 6.348481e-06 0.006549552</span></span>
<span><span class="co">#&gt; ...</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Use $Z.trace for posterior probabilities of topic assignments for each cell</span></span>
<span><span class="co">#&gt; Use $cell_topics for final topic assignments for each cell</span></span>
<span><span class="co">#&gt; Use $parameters for accessing model parameters</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="topic-content-and-distribution">Topic Content and Distribution<a class="anchor" aria-label="anchor" href="#topic-content-and-distribution"></a>
</h3>
<p>SpaTopic identify seven topics from the image. Below we use the
heatmap to show the cell type composition within each topic.</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">pheatmap</span><span class="op">)</span></span>
<span><span class="va">m</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/as.data.frame.html" class="external-link">as.data.frame</a></span><span class="op">(</span><span class="va">gibbs.res</span><span class="op">$</span><span class="va">Beta</span><span class="op">)</span></span>
<span><span class="fu">pheatmap</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/pheatmap/man/pheatmap.html" class="external-link">pheatmap</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/t.html" class="external-link">t</a></span><span class="op">(</span><span class="va">m</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p><img src="SpaTopic_files/figure-html/heatmap-1.png" width="672"></p>
<p>We assign each cell to a topic with the highest posterior probability
and visualize the distribution of cell topics over the image.</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## you can can access the posterior probabilities of topic assignments for each cell</span></span>
<span><span class="va">prob</span><span class="op">&lt;-</span><span class="fu"><a href="https://rdrr.io/r/base/matrix.html" class="external-link">as.matrix</a></span><span class="op">(</span><span class="va">gibbs.res</span><span class="op">$</span><span class="va">Z.trace</span><span class="op">)</span></span>
<span><span class="va">nano.obj</span><span class="op">$</span><span class="va">Topic</span><span class="op">&lt;-</span><span class="fu"><a href="https://rdrr.io/r/base/factor.html" class="external-link">as.factor</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/apply.html" class="external-link">apply</a></span><span class="op">(</span><span class="va">prob</span>,<span class="fl">1</span>,<span class="va">which.max</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">## In the new version of SpaTopic, the final topic assignments is in cell topics.</span></span>
<span><span class="va">nano.obj</span><span class="op">$</span><span class="va">Topic</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/factor.html" class="external-link">as.factor</a></span><span class="op">(</span><span class="va">gibbs.res</span><span class="op">$</span><span class="va">cell_topics</span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://ggplot2.tidyverse.org" class="external-link">ggplot2</a></span><span class="op">)</span></span>
<span><span class="va">palatte</span><span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"#0000FFFF"</span>,<span class="st">"#FF0000FF"</span>,<span class="st">"#00FF00FF"</span>,<span class="st">"#009FFFFF"</span>,<span class="st">"#ff00b7fa"</span>,<span class="st">"#005300FF"</span>,<span class="st">"#FFD300FF"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://satijalab.org/seurat/reference/ImageDimPlot.html" class="external-link">ImageDimPlot</a></span><span class="op">(</span><span class="va">nano.obj</span>, fov <span class="op">=</span> <span class="st">"lung5.rep1"</span>, group.by <span class="op">=</span> <span class="st">"Topic"</span>, axes <span class="op">=</span> <span class="cn">TRUE</span>, </span>
<span>                           dark.background <span class="op">=</span> <span class="cn">T</span>,cols <span class="op">=</span> <span class="va">palatte</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">ggtitle</a></span><span class="op">(</span><span class="st">"Topic"</span><span class="op">)</span> </span></code></pre></div>
<p><img src="SpaTopic_files/figure-html/topics-1.png" width="672"></p>
</div>
<div class="section level3">
<h3 id="compare-to-buildnicheassay-in-seurat-v5">Compare to BuildNicheAssay() in Seurat v5<a class="anchor" aria-label="anchor" href="#compare-to-buildnicheassay-in-seurat-v5"></a>
</h3>
<p>We compare <code>SpaTopic</code> to the function
<code><a href="https://satijalab.org/seurat/reference/BuildNicheAssay.html" class="external-link">BuildNicheAssay()</a></code> in Seurat v5. It took around 5 min on the
same laptop.</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">### NOT RUN!! We use the pre-computed result</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/system.time.html" class="external-link">system.time</a></span><span class="op">(</span><span class="va">nano.obj</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/BuildNicheAssay.html" class="external-link">BuildNicheAssay</a></span><span class="op">(</span>object <span class="op">=</span> <span class="va">nano.obj</span>, <span class="st">"lung5.rep1"</span>, group.by <span class="op">=</span> <span class="st">"predicted.annotation.l1"</span>,</span>
<span>                            niches.k <span class="op">=</span> <span class="fl">7</span>, neighbors.k <span class="op">=</span> <span class="fl">100</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p>We also visualize the distribution of seven niches over the same
image.</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">nano.obj</span><span class="op">$</span><span class="va">niches</span><span class="op">&lt;-</span><span class="fu"><a href="https://rdrr.io/r/base/factor.html" class="external-link">factor</a></span><span class="op">(</span><span class="va">nano.obj</span><span class="op">$</span><span class="va">niches</span><span class="op">)</span></span>
<span><span class="va">nano.obj</span><span class="op">$</span><span class="va">niches</span><span class="op">&lt;-</span><span class="fu"><a href="https://rdrr.io/r/base/factor.html" class="external-link">ordered</a></span><span class="op">(</span><span class="va">nano.obj</span><span class="op">$</span><span class="va">niches</span>,levels <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span>,<span class="fl">2</span>,<span class="fl">3</span>,<span class="fl">4</span>,<span class="fl">5</span>,<span class="fl">6</span>,<span class="fl">7</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">## try to match the colors of topics</span></span>
<span><span class="va">palatte2</span><span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"#FF00B6FF"</span>,<span class="st">"#0000FFFF"</span>,<span class="st">"#FFD300FF"</span>,<span class="st">"#009FFFFF"</span>,<span class="st">"#FF0000FF"</span>,<span class="st">"#005300FF"</span>,<span class="st">"#00FF00FF"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://satijalab.org/seurat/reference/ImageDimPlot.html" class="external-link">ImageDimPlot</a></span><span class="op">(</span><span class="va">nano.obj</span>, fov <span class="op">=</span> <span class="st">"lung5.rep1"</span>, group.by <span class="op">=</span> <span class="st">"niches"</span>, axes <span class="op">=</span> <span class="cn">TRUE</span>, dark.background <span class="op">=</span> <span class="cn">T</span>,cols <span class="op">=</span> <span class="va">palatte2</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">ggtitle</a></span><span class="op">(</span><span class="st">"Niches"</span><span class="op">)</span> </span></code></pre></div>
<p><img src="SpaTopic_files/figure-html/niches_vis-1.png" width="672"></p>
</div>
<div class="section level3">
<h3 id="visualizing-topic-probability-distributions">Visualizing Topic Probability Distributions<a class="anchor" aria-label="anchor" href="#visualizing-topic-probability-distributions"></a>
</h3>
<p>After running SpaTopic, users can visualize not just the discrete
topic assignments but also the probability distribution of each topic
across the tissue.</p>
<p>Here’s how to create these visualizations:</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Extract topic probabilities for each cell and add as metadata to the Seurat object</span></span>
<span><span class="va">topic_prob</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/matrix.html" class="external-link">as.matrix</a></span><span class="op">(</span><span class="va">gibbs.res</span><span class="op">$</span><span class="va">Z.trace</span><span class="op">)</span></span>
<span><span class="kw">for</span><span class="op">(</span><span class="va">i</span> <span class="kw">in</span> <span class="fl">1</span><span class="op">:</span><span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">ncol</a></span><span class="op">(</span><span class="va">topic_prob</span><span class="op">)</span><span class="op">)</span> <span class="op">{</span></span>
<span>  <span class="va">nano.obj</span><span class="op">[[</span><span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="st">"Topic"</span>, <span class="va">i</span>, <span class="st">"_probability"</span><span class="op">)</span><span class="op">]</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="va">topic_prob</span><span class="op">[</span>,<span class="va">i</span><span class="op">]</span></span>
<span><span class="op">}</span></span>
<span></span>
<span><span class="co"># Function to create a topic probability plot</span></span>
<span><span class="va">plot_topic_prob</span> <span class="op">&lt;-</span> <span class="kw">function</span><span class="op">(</span><span class="va">topic_num</span><span class="op">)</span> <span class="op">{</span></span>
<span>  <span class="va">feature_name</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="st">"Topic"</span>, <span class="va">topic_num</span>, <span class="st">"_probability"</span><span class="op">)</span></span>
<span>  <span class="fu"><a href="https://satijalab.org/seurat/reference/ImageFeaturePlot.html" class="external-link">ImageFeaturePlot</a></span><span class="op">(</span><span class="va">nano.obj</span>,</span>
<span>                  fov <span class="op">=</span> <span class="st">"lung5.rep1"</span>,</span>
<span>                  features <span class="op">=</span> <span class="va">feature_name</span>,</span>
<span>                  axes <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>                  dark.background <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>                  cols <span class="op">=</span> <span class="fu">viridis</span><span class="fu">::</span><span class="fu"><a href="https://sjmgarnier.github.io/viridisLite/reference/viridis.html" class="external-link">magma</a></span><span class="op">(</span><span class="fl">100</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">ggtitle</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="st">"Topic "</span>, <span class="va">topic_num</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html" class="external-link">theme</a></span><span class="op">(</span></span>
<span>      legend.position <span class="op">=</span> <span class="st">"right"</span>,</span>
<span>      plot.title <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">element_text</a></span><span class="op">(</span>size <span class="op">=</span> <span class="fl">10</span><span class="op">)</span>,</span>
<span>      plot.margin <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">margin</a></span><span class="op">(</span><span class="fl">1</span>, <span class="fl">1</span>, <span class="fl">1</span>, <span class="fl">1</span>, <span class="st">"pt"</span><span class="op">)</span>  <span class="co"># Reduce margins (top, right, bottom, left)</span></span>
<span>    <span class="op">)</span></span>
<span><span class="op">}</span></span>
<span></span>
<span><span class="co"># Create plots for each topic</span></span>
<span><span class="va">topic_plots</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/lapply.html" class="external-link">lapply</a></span><span class="op">(</span><span class="fl">1</span><span class="op">:</span><span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">ncol</a></span><span class="op">(</span><span class="va">topic_prob</span><span class="op">)</span>, <span class="va">plot_topic_prob</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Combine into a multi-panel figure</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://patchwork.data-imaginist.com" class="external-link">patchwork</a></span><span class="op">)</span></span>
<span><span class="fu"><a href="https://patchwork.data-imaginist.com/reference/wrap_plots.html" class="external-link">wrap_plots</a></span><span class="op">(</span><span class="va">topic_plots</span>, ncol <span class="op">=</span> <span class="fl">2</span><span class="op">)</span> <span class="op">&amp;</span> </span>
<span>  <span class="fu"><a href="https://patchwork.data-imaginist.com/reference/plot_layout.html" class="external-link">plot_layout</a></span><span class="op">(</span>guides <span class="op">=</span> <span class="st">"collect"</span><span class="op">)</span> </span></code></pre></div>
<p><img src="SpaTopic_files/figure-html/topic_prob-1.png" width="960"></p>
</div>
</div>
<div class="section level2">
<h2 id="caf-spatial-subtypes-across-different-topics">CAF spatial subtypes across different topics<a class="anchor" aria-label="anchor" href="#caf-spatial-subtypes-across-different-topics"></a>
</h2>
<p>Inspured by <a href="https://www.cell.com/cancer-cell/fulltext/S1535-6108(25)00083-2" class="external-link">Liu
et al 2025</a>, SpaTopic can also help to identify CAF spatial subtypes.
Specifically:</p>
<ul>
<li>Topic 6 is associated with s1-CAFs adjacent to the tumor;</li>
<li>Topics 5 and 7 correspond to s2-CAFs within the stromal niche;</li>
<li>Topic 1 is linked to s3-CAFs in the myeloid niche;</li>
<li>and Topic 3, primarily comprising B cells and CD4 T cells along with
some myofibroblasts, corresponds to s4-CAFs in the TLS niche.</li>
</ul>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://satijalab.github.io/seurat-object/reference/DefaultAssay.html" class="external-link">DefaultAssay</a></span><span class="op">(</span><span class="va">nano.obj</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="st">"Nanostring"</span></span>
<span><span class="fu"><a href="https://satijalab.github.io/seurat-object/reference/Idents.html" class="external-link">Idents</a></span><span class="op">(</span><span class="va">nano.obj</span><span class="op">)</span><span class="op">&lt;-</span><span class="st">"predicted.annotation.l1"</span></span>
<span><span class="co">## gene list from [Liu et al 2025]</span></span>
<span><span class="va">caf_genes</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"TPM2"</span>, <span class="st">"MYH11"</span>, <span class="st">"ACTA2"</span>, <span class="st">"COL1A1"</span>, <span class="st">"TGFB1"</span>, <span class="st">"LIF"</span>, </span>
<span>               <span class="st">"COL12A1"</span>, <span class="st">"IL6"</span>, <span class="st">"HSPA1A"</span>, <span class="st">"CFD"</span>, <span class="st">"CXCL2"</span>, <span class="st">"ENG"</span>, </span>
<span>               <span class="st">"C1QB"</span>, <span class="st">"CD74"</span>, <span class="st">"CXCL12"</span>, <span class="st">"HLA-DPB1"</span>, <span class="st">"CCL8"</span>, <span class="st">"TAP2"</span><span class="op">)</span></span>
<span></span>
<span><span class="co">### we only look at CAF</span></span>
<span><span class="va">myofibro_obj</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/subset.html" class="external-link">subset</a></span><span class="op">(</span><span class="va">nano.obj</span>, <span class="va">predicted.annotation.l1</span> <span class="op"><a href="https://rdrr.io/r/base/match.html" class="external-link">%in%</a></span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"Myofibroblast"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">selected_topics</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"1"</span>, <span class="st">"3"</span>, <span class="st">"5"</span>, <span class="st">"6"</span>, <span class="st">"7"</span><span class="op">)</span></span>
<span><span class="va">myofibro_filtered</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/subset.html" class="external-link">subset</a></span><span class="op">(</span><span class="va">myofibro_obj</span>, <span class="va">Topic</span> <span class="op"><a href="https://rdrr.io/r/base/match.html" class="external-link">%in%</a></span> <span class="va">selected_topics</span><span class="op">)</span></span>
<span></span>
<span><span class="va">custom_topic_labels</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span></span>
<span>  <span class="st">"6"</span> <span class="op">=</span> <span class="st">"Topic6 - s1-CAFs (Tumor adjacent)"</span>,</span>
<span>  <span class="st">"7"</span> <span class="op">=</span> <span class="st">"Topic7 - s2-CAFs (Stromal niche)"</span>,</span>
<span>  <span class="st">"5"</span> <span class="op">=</span> <span class="st">"Topic5 - s2-CAFs (Stromal niche)"</span>,</span>
<span>  <span class="st">"1"</span> <span class="op">=</span> <span class="st">"Topic1 - s3-CAFs (Myeloid niche)"</span>,</span>
<span>  <span class="st">"3"</span> <span class="op">=</span> <span class="st">"Topic3 - s4-CAFs (TLS niche)"</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://ggplot2.tidyverse.org" class="external-link">ggplot2</a></span><span class="op">)</span></span>
<span><span class="va">red_blue_colors</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/grDevices/colorRamp.html" class="external-link">colorRampPalette</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"blue"</span>, <span class="st">"white"</span>, <span class="st">"red"</span><span class="op">)</span><span class="op">)</span><span class="op">(</span><span class="fl">100</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://satijalab.org/seurat/reference/DotPlot.html" class="external-link">DotPlot</a></span><span class="op">(</span><span class="va">myofibro_filtered</span>, features <span class="op">=</span> <span class="va">caf_genes</span>, group.by <span class="op">=</span> <span class="st">"Topic"</span><span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_discrete.html" class="external-link">scale_y_discrete</a></span><span class="op">(</span>labels <span class="op">=</span> <span class="va">custom_topic_labels</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_gradient.html" class="external-link">scale_color_gradientn</a></span><span class="op">(</span>colors <span class="op">=</span> <span class="va">red_blue_colors</span><span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="fu"><a href="https://satijalab.org/seurat/reference/SeuratTheme.html" class="external-link">RotatedAxis</a></span><span class="op">(</span><span class="op">)</span><span class="op">+</span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_size.html" class="external-link">scale_size_continuous</a></span><span class="op">(</span></span>
<span>    range <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">0</span>, <span class="fl">8</span><span class="op">)</span>,  <span class="co"># Min and max dot size</span></span>
<span>    breaks <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">0</span>, <span class="fl">10</span>, <span class="fl">20</span>, <span class="fl">30</span>, <span class="fl">40</span>, <span class="fl">50</span>, <span class="fl">60</span>, <span class="fl">70</span><span class="op">)</span>,  <span class="co"># Custom breaks</span></span>
<span>    limits <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">0</span>, <span class="fl">100</span><span class="op">)</span>  <span class="co"># Set limits to ensure all breaks are included</span></span>
<span>  <span class="op">)</span> </span></code></pre></div>
<p><img src="SpaTopic_files/figure-html/cafsubtype-1.png" width="960"></p>
<p>Though CAF in stroma niche generally have common gene expression
profile, there are some CAF genes that are differentially expressed
between Topic 5 and Topic 7, such as TGFB1, HSP1A1, COL12A1 genes,
potentially due to different immune-enriched neighborhood in the stroma
region.</p>
</div>
<div class="section level2">
<h2 id="topic-inference-on-multiple-images">Topic Inference on Multiple Images<a class="anchor" aria-label="anchor" href="#topic-inference-on-multiple-images"></a>
</h2>
<p>SpaTopic can identify common tissue patterns across multiple images.
The input should be a list of data frames. See an example below (not
run).</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## tissue1, tissue2 are data frames of two different images.</span></span>
<span><span class="va">gibbs.res</span><span class="op">&lt;-</span><span class="fu"><a href="../reference/SpaTopic_inference.html">SpaTopic_inference</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>A <span class="op">=</span> <span class="va">tissue1</span>, B <span class="op">=</span> <span class="va">tissue2</span><span class="op">)</span>, ntopics <span class="op">=</span> <span class="fl">7</span>, sigma <span class="op">=</span> <span class="fl">50</span>, region_radius <span class="op">=</span> <span class="fl">400</span><span class="op">)</span> </span></code></pre></div>
<p>Please check more examples in <a href="https://xiyupeng.github.io/SpaTopic/">SpaTopic Home Page</a>.</p>
</div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p>Developed by Xiyu Peng.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.2.</p>
</div>

    </footer>
</div>





  </body>
</html>
